### ISSUE TITLE
 [✨ feature request]Does the differentiable rendering module support `roughconductor`?

### ISSUE THREAD

**Issue by HsiangYangChu** [mitsuba-renderer/mitsuba2#575]
My question is that does the differentiable rendering module support `roughconductor`? If it does, could you provide an example?
Looking forward to your reply!

**Comment by Speierers** [comment]
What are the parameters that you are willing to optimize? You should be able to optimize the  `specular_reflectance` without any issue, however, for the `eta` and `alpha` parameters, this gets trickier as it will introduce discontinuities in the rendering integral. See [this paper](https://rgl.epfl.ch/publications/Zeltner2021MonteCarlo) for more details.

**Comment by HsiangYangChu** [comment]
Thanks for your reply @Speierers !
I just wanna optimize the `specular_reflectance`. What should I do? Should I give every vertex a `specular_reflenctance` variable just like `xyzrgb`?

**Comment by Speierers** [comment]
You can texture this parameter using a `Bitmap` plugin for instance, and then optimize the texture data scene parameter in the same way we optimize the albedo of the cbox wall in [this tutorial](https://mitsuba2.readthedocs.io/en/latest/src/inverse_rendering/diff_render.html).

**Comment by HsiangYangChu** [comment]
Thanks!
I will have a try!

**Comment by HsiangYangChu** [comment]
Hi @Speierers
I have a try as you said. But I get a bug as follows:
```
  what():  ​[xml.cpp:1031] Error while loading "pig/render.xml" (near line 13, col 4): unreferenced object BitmapTextureImpl[
[xml.cpp:1031]   name = "roughness.exr",
[xml.cpp:1031]   resolution = "[800, 800]",
[xml.cpp:1031]   raw = 0,
[xml.cpp:1031]   mean = 0.0866291,
[xml.cpp:1031]   transform = [[1, 0, 0],
[xml.cpp:1031]    [0, 1, 0],
[xml.cpp:1031]    [0, 0, 1]]
[xml.cpp:1031] ] (within bsdf of type "roughconductor")
Aborted (core dumped)
```
This is my `xml` code:
```
            <shape type="ply">
                <string name="filename" value="{geo_path}"/>
                <bsdf type="roughconductor">
                    <string name="distribution" value="ggx"/>
                    <texture type="bitmap" name="diffuseReflectance">
				        <string name="filename" value="diffuse.exr"/>
	            </texture>
                </bsdf>
            </shape>
```
Besides, Is the `ward` model be supported by Mitsuba2.0?

Looking forward to your reply!

**Comment by Speierers** [comment]
Looks like the name used for the texture in the xml isn't right.

No the ward model isn't implemented but it shouldn't be too hard to port from Mitsuba 0.6

**Comment by merlinND** [comment]
@HsiangYangChu An easy way to find the property names used by a plugin is to check in the constructor. In your case:

https://github.com/mitsuba-renderer/mitsuba2/blob/4e7628c6eed365904ca2ba536b795d1b03410344/src/bsdfs/roughconductor.cpp#L185-L186

**Comment by HsiangYangChu** [comment]
> @HsiangYangChu An easy way to find the property names used by a plugin is to check in the constructor. In your case:
> 
> https://github.com/mitsuba-renderer/mitsuba2/blob/4e7628c6eed365904ca2ba536b795d1b03410344/src/bsdfs/roughconductor.cpp#L185-L186

Awesome! I appreciate it!

**Comment by HsiangYangChu** [comment]
> @HsiangYangChu An easy way to find the property names used by a plugin is to check in the constructor. In your case:
> 
> https://github.com/mitsuba-renderer/mitsuba2/blob/4e7628c6eed365904ca2ba536b795d1b03410344/src/bsdfs/roughconductor.cpp#L185-L186

Awesome! I appreciate it!

**Comment by HsiangYangChu** [comment]
Hi @merlinND,
I succeeded! Thanks again!
Now, I want to optimize `diffuse` and `specular`, what brdf type should I choose?

**Comment by Speierers** [comment]
The current codebase doesn't provide any material with both a specular and diffuse lobe that's "optimizable". You could try to blend two BSDF with the `blend` BSDF.

Otherwise you can wait until the next release which will  provide such model.

Good luck!

**Comment by HsiangYangChu** [comment]
@Speierers, Hi
I succeeded in reconstructing the specular and diffuse parameters! Thanks again!
I use `mse` loss function, but I find it not satisfying, So, I want to constrain the specular and diffuse parameters separately. 
So, Does `mitsuba2`  support Regularization?
Looking forward to you reply!

**Comment by Speierers** [comment]
Yes it is up to you to define you loss function in Python. Thanks to the AD capability of Enoki, you can perform some  differentiable computation on the parameters themselves and add this term to the loss before backpropagating.

**Comment by HsiangYangChu** [comment]
I really appreciate it!!!!

**Comment by HsiangYangChu** [comment]
Sorry to bother you again!
What is the shading frequency used in this renderer? (Flat shading, Gouraud shading, or Phong shading)
Looking forward to your reply!

**Comment by Speierers** [comment]
Not sure to understand your question. In a physically based renderer the "shading" of an object is defined by it's BSDF. IIRC you were experimenting with the [`roughconductor` BSDF plugin](https://mitsuba2.readthedocs.io/en/latest/generated/plugins.html#rough-conductor-material-roughconductor).

If by "shading" you mean the interpolation scheme used to compute the shading normals, I can tell you that it is based on [this paper](https://www.tandfonline.com/doi/abs/10.1080/10867651.1998.10487487?journalCode=ujgt19).

**Comment by HsiangYangChu** [comment]
Thank you so much! @Speierers
Actually, I mean the interpolation scheme using barycentric coordinate used to compute texture coordinate. 
Concretely, there are two options. One can find the texture of the three vertices according to the texture coordinates of the three vertices, and then get the texture of each point in the triangle by interpolating according to the barycentric coordinates. Alternatively, one can interpolate the texture coordinates of three vertices to obtain the texture coordinates of each point in the triangle, and then find the texture map to obtain the texture of that point

**Comment by Speierers** [comment]
In Mitsuba we interpolate the texture coordinates of the 3 vertices of the hit triangle using baricentric coordinates

**Comment by HsiangYangChu** [comment]
Thanks!!!

**Comment by HsiangYangChu** [comment]
> Yes it is up to you to define you loss function in Python. Thanks to the AD capability of Enoki, you can perform some differentiable computation on the parameters themselves and add this term to the loss before backpropagating.

I am trying to do this. The following is my test code:
![image](https://user-images.githubusercontent.com/35549898/162347593-a588118d-b5c6-4bfb-8254-3a60ab4b2f5c.png)
But I found it does not converge! Did I do something wrong?

**Comment by Speierers** [comment]
For such simple code I would stick with enoki and avoid going to the PyTorch realm. I suppose in your case the gradients don't properly propagate from PyTorch to the Enoki variable.

Unfortunately you will be on your own to debug this as we are focusing on the upcoming release.

**Comment by HsiangYangChu** [comment]
Thanks!!!
When will the next version be released? I am really looking forward to that! :)


| |
朱向阳
|
|
***@***.***
|
签名由网易邮箱大师定制


On 04/8/2022 14:39，Sebastien ***@***.***> wrote：

For such simple code I would stick with enoki and avoid going to the PyTorch realm. I suppose in your case the gradients don't properly propagate from PyTorch to the Enoki variable.

Unfortunately you will be on your own to debug this as we are focusing on the upcoming release.
